File: examples/0_basic/4_delete_did.rs
// Copyright 2020-2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use examples::get_funded_client;
use examples::get_memstorage;
use identity_iota::iota::rebased::Error;
use identity_iota::iota::IotaDocument;

/// Demonstrates how to delete a DID in an identity.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
  let storage = get_memstorage()?;
  let client = get_funded_client(&storage).await?;
  let mut identity = client
    .create_identity(IotaDocument::new(client.network()))
    .finish()
    .build_and_execute(&client)
    .await?
    .output;
  let did = identity.did_document().id().clone();

  println!("Created a new Identity containing DID Document {did}");

  // Delete the DID we just created.
  let controller_token = identity.get_controller_token(&client).await?.expect("is a controller");
  identity
    .delete_did(&controller_token)
    .finish(&client)
    .await?
    .build_and_execute(&client)
    .await?;

  assert!(identity.has_deleted_did());
  assert_eq!(identity.did_document().metadata.deactivated, Some(true));

  println!("DID {did} was successfully deleted!");

  // Trying to update a deleted DID Document must fail.
  let err = identity
    .update_did_document(IotaDocument::new(client.network()), &controller_token)
    .finish(&client)
    .await;
  assert!(matches!(err, Err(Error::Identity(_))));

  // Resolution of the DID document through its DID must fail.
  let err = client.resolve_did(&did).await.unwrap_err();
  assert!(matches!(err, Error::DIDResolutionError(_)));

  Ok(())
}

File: bindings/wasm/identity_wasm/examples/src/0_basic/4_delete_did.ts
// Copyright 2020-2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { IotaClient } from "@iota/iota-sdk/client";
import { createDocumentForNetwork, getFundedClient, getMemstorage, NETWORK_URL, TEST_GAS_BUDGET } from "../util";

/** Demonstrates how to delete a DID of an identity. */
export async function deleteIdentityDID() {
    // create new clients and create new account
    const iotaClient = new IotaClient({ url: NETWORK_URL });
    const network = await iotaClient.getChainIdentifier();
    const storage = getMemstorage();
    const identityClient = await getFundedClient(storage);
    const [unpublished] = await createDocumentForNetwork(storage, network);

    // create new identity for this account and publish document for it
    const { output: identity } = await identityClient
        .createIdentity(unpublished)
        .finish()
        .buildAndExecute(identityClient);
    const did = identity.didDocument().id();

    // Resolve the latest state of the document.
    // Technically this is equivalent to the document above.
    const resolved = await identityClient.resolveDid(did);
    console.log("Resolved DID document:", JSON.stringify(resolved, null, 2));

    const controllerToken = await identity.getControllerToken(identityClient);

    // delete the DID.
    await identity
        .deleteDid(controllerToken!)
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(identityClient);

    // After an Identity's DID has been deleted, the document will be
    // empty and inactive. Identity.hasDeletedDid must return `true`.
    const is_deleted = identity.didDocument().metadata().deactivated()
        && identity.hasDeletedDid();
    if (!is_deleted) {
        throw new Error("failed to delete DID Document");
    }

    // Resolving a deleted DID must throw an error.
    try {
        let deactivated = await identityClient.resolveDid(did);
    } catch (_) {
        console.log(`DID ${did} was successfully deleted!`);
    }

    // Trying to update a deleted DID must fail!
    try {
        await identity
            .updateDidDocument(resolved, controllerToken!)
            .withGasBudget(TEST_GAS_BUDGET)
            .buildAndExecute(identityClient);
    } catch (_) {
        console.log("A deleted DID cannot be updated!");
    }
}
